### DISCUSSION TITLE
Implicit Settings on Button Callback

### DISCUSSION THREAD

**Discussion by jdbodyfelt** [upgradeQ/Streaming-Software-Scripting-Reference#24] [Q&A]
Hi, sorry to bother it's me again. I have another question popping up in the `property_modification.py` example. Namely, in `callback` you mention that settings are passed implicitly, however, how do you access them within the callback? I've tried a `for key, val in kwargs.items(): print(key, val)`, but it does not produce anything. Thank you!

**Comment by jdbodyfelt** [comment]
As an update, I can see that `args[0]` holds an `obs_data_t` object initially (assuming it's `settings`), but after pushing the button, it's  emptied (as well as `kwargs`).

**Comment by upgradeQ** [comment]
What are you trying to do? Have you found a [post](https://obsproject.com/forum/threads/tips-and-tricks-for-lua-scripts.132256/#post-487388)  from @bfxdev about properties ? Maybe this is what are you looking for. 
![img](https://raw.githubusercontent.com/bfxdev/OBS/master/pics/visibility.gif)
Also a [page](https://github.com/obsproject/obs-studio/wiki/Getting-Started-With-OBS-Scripting) on the wiki

---

The `property_modification.py` example snippet was written in June 2020, and there was no intention to access settings in callback. They are passed implicitly using star arguments because later you cannot access it ( as you confirmed it ). `**kwargs` were passed just to make sure nothing breaks. 
Take a look to a new example:
```python
import obspython as obs
from itertools import count

G = lambda: ...
G.settings = None
G.n = 0
G.counter = count()


def refresh_pressed(props, prop, *args):  # pass settings implicitly
    print("refresh pressed")
    if args:
        G.counter = count(obs.obs_data_get_int(args[0], "saved_int"))
    G.n = next(G.counter)
    p = obs.obs_properties_get(props, "button")
    obs.obs_property_set_description(p, f"refresh pressed {G.n} times")
    return True


def script_description():
    return "Modify property and save/load from settings "


def script_load(settings):
    G.settings = settings


def script_save(settings):
    obs.obs_data_set_int(settings, "saved_int", G.n)
    G.settings = settings


def script_properties():
    props = obs.obs_properties_create()
    b = obs.obs_properties_add_button(
        props, "button", "refresh pressed m times", refresh_pressed
    )
    obs.obs_property_set_modified_callback(b, refresh_pressed)
    obs.obs_properties_apply_settings(props, G.settings)
    return props
```
In order to access settings in callback they must be global.

**Comment by jdbodyfelt** [comment]
Thank you! On button press, I'm looking to auto-iterate integer counter widgets. Been using the global solution, but seems in `script_update` I may have been assigning in the wrong direction (`settings = G.settings`). I'll play some more with it, and if I get it working I'll post my solution...

**Comment by MarioMey** [comment]
@jdbodyfelt if you have more questions, ask them in [Discussion section](https://github.com/upgradeQ/OBS-Studio-Python-Scripting-Cheatsheet-obspython-Examples-of-API/discussions). There is where I ask my doubts 😉.
